src/plugins/data/public/search/session_service.ts

```diff
@@ -0,0 +1,141 @@
+    this.idMapping = new Map();
+    this.start();
+  }
+
+  public trackSearch(request: IEsSearchRequest, sessionId: string | undefined) {
+    if (sessionId && request.params && request.params.body) {
+      const sessionInfo = this.idMapping.get(sessionId);
+
+      // Mark a session as running.
+      // Also reopen a complete session, if a new search is run. We know this can happen because of follow up requests.
+      if (
+        sessionInfo!.status === SessionStatus.New ||
+        sessionInfo!.status === SessionStatus.Completed
+      ) {
+        sessionInfo!.status = SessionStatus.Running;
+      }
+
+      // Add request info to the session
+      sessionInfo!.requests[createRequestHash(request.params.body)] = {
+        status: SearchStatus.Running,
+      };
+    }
+  }
+
+  public trackSearchId(request: IEsSearchRequest, sessionId: string | undefined, searchId: string) {
+    if (sessionId && request.params && request.params.body) {
+      const sessionInfo = this.idMapping.get(sessionId);
+      const requestInfo = sessionInfo!.requests[createRequestHash(request.params.body)];
+      if (requestInfo) {
+        requestInfo.searchId = searchId;
+      }
+    }
+  }
+
+  public trackSearchComplete(request: IEsSearchRequest, sessionId?: string) {
+    if (sessionId && request.params && request.params.body) {
+      const sessionInfo = this.idMapping.get(sessionId);
+      sessionInfo!.requests[createRequestHash(request.params.body)].status = SearchStatus.Done;
+
+      // Mark session as done, if all requests are done
+      Object.values(sessionInfo!.requests)
+        .map((requestInfo) => requestInfo.status === SearchStatus.Done)
+        .every(() => {
+          sessionInfo!.status = SessionStatus.Completed;
+        });
+    }
+  }
+
+  public trackSearchError(request: IEsSearchRequest, sessionId?: string, e?: Error) {
+    if (sessionId && request.params && request.params.body) {
+      const sessionInfo = this.idMapping.get(sessionId);
+
+      // Mark request as errored, don't update session status
+      if (e instanceof SearchTimeoutError) {
+        sessionInfo!.status = SessionStatus.Timeout;
+      }
+      sessionInfo!.requests[createRequestHash(request.params.body)].status = SearchStatus.Error;
+    }
+  }
+
+  public getSessionId() {
+    return this.sessionId;
+  }
+
+  public getSessionTimeoutNotified(): boolean {
+    if (this.sessionId) {
+      const sessionInfo = this.idMapping.get(this.sessionId);
+      return sessionInfo!.timeoutNotified;
+    } else {
+      return false;
+    }
+  }
+
+  public setSessionTimeoutNotified() {
+    if (this.sessionId) {
+      const sessionInfo = this.idMapping.get(this.sessionId);
+      sessionInfo!.timeoutNotified = true;
+    }
+  }
+
+  public start() {
+    this.sessionId = uuid.v4();
+
+    const sessionInfo = {
+      requests: {},
+      status: SessionStatus.New,
+      timeoutNotified: false,
+    };
+    this.idMapping.set(this.sessionId, sessionInfo);
```

Comment:
I worry that we are never clearing this. For a dashboard on a screen (like a fixed monitor) that never gets refreshed, this will just continue to chew up memory. I wonder what we can do to clear this out from time to time without losing the functionality we want.

Reply:
I added a clear upon navigation (inside the clear method).
What do you think?

Reply:
i guess this still doesn't address the scenario lukas mentioned, with dashboard running on a screen for ever. If memory continues to grow this will lead to a crash after a certain amount of time, which i don't think is acceptable.

Reply:
After talking to @ppisljar I changed the FE service to have a single active session for now.